cpp/src/arrow/type_fwd.h

```diff
@@ -89,6 +89,9 @@ using ChunkedArrayVector = std::vector<std::shared_ptr<ChunkedArray>>;
 using RecordBatchVector = std::vector<std::shared_ptr<RecordBatch>>;
 using RecordBatchIterator = Iterator<std::shared_ptr<RecordBatch>>;
 
+template <typename TypeClass>
+class TypedArrayStatistics;
```

Comment:
A new typed class hierarchy for statistics seems like an overkill. Think of statistics as a dynamic object like `ArrayData` and the classes like `BooleanArray`, `StringArray` have typed accessors to the underlying untyped statistics object just like they have typed accessors to the underlying `ArrayData` for the array buffers of that type.

Reply:
And compute kernels that deal directly with untyped ArrayData would deal directly with untyped `ArrayStats` objects.

Reply:
I think that the typed classes are convenient but they are required in the first version. I'll remove them. If we think that they are convenient, we can revisit this later.

Reply:
Removed.